src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -7130,24 +7140,187 @@ public override IOperation VisitArgument(IArgumentOperation operation, int? capt
+                    : new InvocationOperation(operation.CloneMethod, visitedInstance,
+                        isVirtual: true, arguments: ImmutableArray<IArgumentOperation>.Empty,
+                        semanticModel: null, operation.Syntax, operation.Type, isImplicit: true);
+            }
+
+            return PopStackFrame(frame, HandleObjectOrCollectionInitializer(operation.Initializer, cloned));
+
+            // For `old with { Property = ... }` we're going to do the same as `new { Property = ..., OtherProperty = old.OtherProperty }`
+            IOperation handleAnonymousTypeWithExpression(WithOperation operation, int? captureIdForResult)
+            {
+                Debug.Assert(operation.Type!.IsAnonymousType);
+                SpillEvalStack(); // before entering a new region, we ensure that anything that needs spilling was spilled
+
+                // The outer region holds captures for all the values for the anonymous object creation
+                var outerCaptureRegion = CurrentRegionRequired;
+
+                // The inner region holds the capture for the operand (ie. old value)
+                var innerCaptureRegion = new RegionBuilder(ControlFlowRegionKind.LocalLifetime);
+                EnterRegion(innerCaptureRegion);
+
+                var initializers = operation.Initializer.Initializers;
+
+                var properties = operation.Type.GetMembers()
+                    .Where(m => m.Kind == SymbolKind.Property)
+                    .Select(m => (IPropertySymbol)m);
+
+                int oldValueCaptureId;
+                if (setsAllProperties(initializers, properties))
+                {
+                    // Avoid capturing the old value since we won't need it
+                    oldValueCaptureId = -1;
+                    AddStatement(VisitRequired(operation.Operand));
+                }
+                else
+                {
+                    oldValueCaptureId = GetNextCaptureId(innerCaptureRegion);
+                    VisitAndCapture(operation.Operand, oldValueCaptureId);
+                }
+                // calls to Visit may enter regions, so we reset things
+                LeaveRegionsUpTo(innerCaptureRegion);
+
+                var explicitProperties = new Dictionary<IPropertySymbol, IOperation>(SymbolEqualityComparer.IgnoreAll);
+                var initializerBuilder = ArrayBuilder<IOperation>.GetInstance(initializers.Length);
+
+                // Visit and capture all the values, and construct assignments using capture references
+                foreach (IOperation initializer in initializers)
+                {
+                    if (initializer is not ISimpleAssignmentOperation simpleAssignment)
+                    {
+                        AddStatement(VisitRequired(initializer));
+                        continue;
+                    }
+
+                    if (simpleAssignment.Target is InvalidOperation)
```

Comment:
>InvalidOperation [](http://example.com/codeflow?start=51&length=16)

Since below we are casting to IPropertyReferenceOperation, consider performing a type check for that instead. If the only alternative that we expect is an InvalidOperation, we can assert that when the type check fails.